pykafka/managedbalancedconsumer.py

```diff
@@ -0,0 +1,368 @@
+
+        self._rebalancing_lock = cluster.handler.Lock()
+        self._consumer = None
+        self._consumer_id = b''
+        self._worker_trace_logged = False
+        self._worker_exception = None
+
+        if auto_start is True:
+            self.start()
+
+    def _setup_heartbeat_worker(self):
+        """Start the heartbeat worker"""
+        self = weakref.proxy(self)
+
+        def fetcher():
+            while True:
+                try:
+                    if not self._running:
+                        break
+                    self._send_heartbeat()
+                    self._cluster.handler.sleep(self._heartbeat_interval_ms / 1000)
+                except ReferenceError:
+                    break
+                except Exception:
+                    # surface all exceptions to the main thread
+                    self._worker_exception = sys.exc_info()
+                    break
+            log.debug("Heartbeat worker exiting")
+        log.info("Starting heartbeat worker")
+        return self._cluster.handler.spawn(
+            fetcher, name="pykafka.ManagedBalancedConsumer.heartbeats")
+
+    def start(self):
+        """Start this consumer. Must be called before consume() if `auto_start=False`."""
+        try:
+            self._running = True
+            self._group_coordinator = self._cluster.get_group_coordinator(
+                self._consumer_group)
+            self._rebalance()
+            self._setup_heartbeat_worker()
+        except Exception:
+            log.exception("Stopping consumer in response to error")
+            self.stop()
+
+    def stop(self):
+        """Stop this consumer
+
+        Should be called as part of a graceful shutdown
+        """
+        self._running = False
+        if self._consumer is not None:
+            self._consumer.stop()
+        self._group_coordinator.leave_managed_consumer_group(self._consumer_group,
+                                                             self._consumer_id)
+
+    def _send_heartbeat(self):
+        """Send a heartbeat request to the group coordinator and react to the response"""
+        res = self._group_coordinator.group_heartbeat(
+            self._consumer_group, self._generation_id, self._consumer_id)
+        if res.error_code == 0:
+            return
+        log.info("Error code %d encountered on heartbeat." % res.error_code)
+        if res.error_code in (IllegalGeneration.ERROR_CODE,
+                              RebalanceInProgress.ERROR_CODE,
+                              UnknownMemberId.ERROR_CODE):
+            pass
+        elif res.error_code in (GroupCoordinatorNotAvailable.ERROR_CODE,
+                                NotCoordinatorForGroup.ERROR_CODE):
+            self._group_coordinator = self._cluster.get_group_coordinator(
+                self._consumer_group)
+        elif res.error_code == GroupAuthorizationFailed.ERROR_CODE:
+            raise GroupAuthorizationFailed()
```

Comment:
I can see why you might see it as out-of-scope -- we could use some abstractions that may be far reaching. However, I'm proposing changes that are completely isolated to code added by this pull request. My suggestion is to move these if-statements into the response classes that you've added.